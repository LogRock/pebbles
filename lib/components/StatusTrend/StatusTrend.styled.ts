import styled from "styled-components";

import { ParagraphLarge } from "../Typography";
import { StatusTrendProps } from "./StatusTrend.types";

export const StatusTrendWrapper = styled.div<Pick<StatusTrendProps, "variant">>`
  display: flex;
  align-items: flex-start;
  justify-content: center;
  width: max-content;
`;

export const LabelWrapper = styled.div<{ hasSecondaryLabel: Boolean }>`
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  justify-content: flex-start;
  margin-top: ${(props: any) => (props.hasSecondaryLabel ? "-7px" : "-4px")};
  margin-left: ${({ theme }) => theme.spacings.sm};
`;

export const StyledPrimaryLabel = styled(ParagraphLarge)<
  Pick<StatusTrendProps, "variant">
>`
  color: ${({ theme, variant }) => theme.statusTrend.labelColor[variant]};
  font-size: ${({ theme }) => theme.statusTrend.primaryLabel.fontSize};
  font-weight: ${({ theme }) => theme.statusTrend.primaryLabel.fontWeight};
`;

export const StyledSecondaryLabel = styled(ParagraphLarge)`
  margin-top: -2px;

  color: ${({ theme }) => theme.colors.neutral[700]};
  font-size: ${({ theme }) => theme.statusTrend.secondaryLabel.fontSize};
  font-weight: ${({ theme }) => theme.statusTrend.secondaryLabel.fontWeight};
  text-transform: uppercase;
`;
