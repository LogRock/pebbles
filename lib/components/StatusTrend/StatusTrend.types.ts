import { spacingTokens } from "../../types/tokens";

export interface StatusTrendProps {
  direction: "up" | "down" | "neutral";
  variant: "negative" | "positive" | "neutral";
  spaceAfter?: spacingTokens;
}
