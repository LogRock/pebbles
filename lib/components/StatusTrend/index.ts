export { default } from "./StatusTrend";
export type { StatusTrendProps } from "./StatusTrend.types";
