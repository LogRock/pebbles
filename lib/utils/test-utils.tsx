/* eslint-disable import/export */
import React from "react";
import { cleanup, render } from "@testing-library/react";
import { afterEach } from "vitest";
import { ThemeProvider } from "styled-components";
import { light as theme } from "../theme";

afterEach(() => {
  cleanup();
});

const customRender = (ui: React.ReactElement, options = {}) =>
  render(ui, {
    // wrap provider(s) here if needed
    wrapper: ({ children }) => (
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    ),
    ...options,
  });

export * from "@testing-library/react";
export { default as userEvent } from "@testing-library/user-event";
// override render export
export { customRender as render };
