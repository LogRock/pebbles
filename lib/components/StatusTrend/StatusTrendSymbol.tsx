import React, { FC, useContext } from "react";
import { ThemeContext } from "styled-components";

import { StatusTrendProps } from "./StatusTrend.types";

export const StatusTrendSymbol: FC<StatusTrendProps> = ({
  variant,
  direction,
}) => {
  const theme = useContext(ThemeContext);

  return (
    <>
      {direction === "up" && (
        <svg
          width="30"
          height="18"
          viewBox="0 0 30 18"
          fill="none"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16.4444 16.2223H28M28 16.2223V4.66672M28 16.2223L16.4444 4.66672L10.6667 10.4445L2 1.77783"
            stroke={theme.statusTrend.iconColor[variant]}
            strokeWidth="2.67"
            strokeLinecap="round"
            strokeLinejoin="round"
          />
        </svg>
      )}
      {direction === "down" && (
        <svg
          width="30"
          height="18"
          viewBox="0 0 30 18"
          fill="none"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16.4444 1.77783H28M28 1.77783V13.3334M28 1.77783L16.4444 13.3334L10.6667 7.55561L2 16.2223"
            stroke={theme.statusTrend.iconColor[variant]}
            strokeWidth="2.67"
            strokeLinecap="round"
            strokeLinejoin="round"
          />
        </svg>
      )}
      {direction === "neutral" && (
        <svg
          width="30"
          height="4"
          viewBox="0 0 30 4"
          fill="none"
          xmlns="http://www.w3.org/2000/svg"
          style={{ marginTop: "5px" }}
        >
          <path
            d="M28 2H2"
            stroke={theme.statusTrend.iconColor[variant]}
            strokeWidth="2.67"
            strokeLinecap="round"
            strokeLinejoin="round"
          />
        </svg>
      )}
    </>
  );
};
