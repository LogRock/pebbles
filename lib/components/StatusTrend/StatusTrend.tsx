import React, { FC, PropsWithChildren } from "react";

import { StatusTrendWrapper } from "./StatusTrend.styled";
import { StatusTrendProps } from "./StatusTrend.types";
import { StatusTrendSymbol } from "./StatusTrendSymbol";

const StatusTrend: FC<PropsWithChildren<StatusTrendProps>> = ({
  direction = "neutral",
  variant = "neutral",
  children,
  spaceAfter,
  ...props
}) => {
  return (
    <StatusTrendWrapper variant={variant} spaceAfter={spaceAfter} {...props}>
      <StatusTrendSymbol direction={direction} variant={variant} />
      {children}
    </StatusTrendWrapper>
  );
};

export default StatusTrend;
