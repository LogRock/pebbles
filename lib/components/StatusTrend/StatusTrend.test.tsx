import React from "react";

import "@testing-library/jest-dom";
import { render, screen } from "../../utils/test-utils";
import StatusTrend from ".";

describe("StatusTrend", () => {
  it("renders properly", () => {
    const { container } = render(
      <StatusTrend direction="down" variant="negative">
        +10.3%
      </StatusTrend>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it("shows the status indicator", () => {
    render(
      <StatusTrend direction="down" variant="negative">
        +10.3%
      </StatusTrend>
    );
    screen.debug();
    expect(screen.getByText("+10.3%")).toBeInTheDocument();
  });

  it("shows the down svg arrow", () => {
    render(
      <StatusTrend direction="down" variant="negative">
        +10.3%
      </StatusTrend>
    );

    expect(screen.getByTestId("down-arrow")).toBeInTheDocument();
  });

  it("shows the up svg arrow", () => {
    render(
      <StatusTrend direction="up" variant="negative">
        +10.3%
      </StatusTrend>
    );

    expect(screen.getByTestId("up-arrow")).toBeInTheDocument();
  });
});
