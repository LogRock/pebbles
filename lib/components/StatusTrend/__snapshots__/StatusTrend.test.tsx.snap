// Vitest Snapshot v1

exports[`StatusTrend > renders properly 1`] = `
<div
  class="sc-aXZVg cgKssw"
>
  <svg
    data-testid="down-arrow"
    fill="none"
    height="6"
    viewBox="0 0 30 18"
    width="10"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M16.4444 1.77783H28M28 1.77783V13.3334M28 1.77783L16.4444 13.3334L10.6667 7.55561L2 16.2223"
      data-testid="down-arrow-path"
      stroke="#DC2626"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2.67"
    />
  </svg>
  +10.3%
</div>
`;
